nuxeo-runtime/nuxeo-runtime-test/src/main/java/org/nuxeo/runtime/test/NXRuntimeTestCase.java

```diff
@@ -398,6 +411,67 @@ public RuntimeContext deployTestContrib(String bundle, URL contrib) throws Excep
         return ctx;
     }
 
+    @Override
+    public void deployPartial(String name, Set<TargetExtensions> targetExtensions) throws Exception {
+        // Do not install bundle; we only need the Object to list his components
+        Bundle bundle = new BundleImpl(osgi, lookupBundle(name), null);
+
+        RuntimeContext ctx = new OSGiRuntimeContext(runtime, bundle);
+        listBundleComponents(bundle).map(URLStreamRef::new).forEach(component -> {
+            try {
+                this.deployPartialComponent(ctx, targetExtensions, component);
+            } catch (IOException e) {
+                log.warn("PartialBundle: " + name + " failed to load: " + component, e);
+            }
+        });
+    }
+
+    /**
+     * Read a component from his StreamRef and create a new component (suffixed with `-partial`, and the base component
+     * name aliased) with only matching contributions of the extensionPoints parameter.
+     * 
+     * @param ctx RuntimeContext in which the new component will be deployed
+     * @param extensionPoints Set of white listed TargetExtensions
+     * @param component Reference to the original component
+     */
+    protected void deployPartialComponent(RuntimeContext ctx, Set<TargetExtensions> extensionPoints,
+            StreamRef component) throws IOException {
+        RegistrationInfo ri = ((DefaultRuntimeContext) ctx).createRegistrationInfo(component);
+        String name = ri.getName().getName() + "-partial";
+
+        // Flatten Target Extension Points
+        Set<Pair<String, String>> targets = extensionPoints.stream()
+                                                           .map(TargetExtensions::getTargetExtensions)
+                                                           .flatMap(Set::stream)
+                                                           .collect(Collectors.toSet());
+
+        String ext = Arrays.stream(ri.getExtensions())
+                           .filter(e -> targets.contains(
+                                   Pair.of(e.getTargetComponent().getName(), e.getExtensionPoint())))
```

Comment:
We don't use the left or rights parts of the `Pair`. This is the only place where we actually check what's in a `Pair` of the `TargetExtensions` and we only do containment comparison of the full `Pair`, so I believe it would simplify the code if we just stored it as a `String` with value `name + ':' + extension` for example.